I tried g++ 7.3.1 and see the same error:
```
g++ (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
W20200708-114754.cpp:54659:19: error: expected unqualified-id before string constant
 NOOPTIMIZE extern "C" ECL_API size32_t tfB9UB1(ARowBuilder & crSelf, const byte * src) {
                   ^~~
W20200708-114754.cpp:82567:19: error: expected unqualified-id before string constant
 NOOPTIMIZE extern "C" ECL_API size32_t tf7G2F1(ARowBuilder & crSelf, const byte * src) {
                   ^~~
```
Did you use clang ?  I thik we need to change the version test.

--- a/ecl/hqlcpp/hqlcatom.cpp
+++ b/ecl/hqlcpp/hqlcatom.cpp
@@ -52,6 +52,8 @@ IAtom * dictionaryAtom;
 IAtom * eclAtom;
 IAtom * endAtom;
 IAtom * ensureCapacityAtom;
+IAtom * externAtom;
+IAtom * externCAtom;
 IAtom * fileAtom;
 IAtom * funnelAtom;
 IAtom * forceAllCheckAtom;
@@ -1469,6 +1471,8 @@ MODULE_INIT(INIT_PRIORITY_HQLATOM-1)
     MAKEATOM(ecl);
     MAKEATOM(end);
     MAKEATOM(ensureCapacity);
+    MAKEATOM(extern);
+    MAKEATOM(externC);
     MAKEATOM(file);
     MAKEATOM(forceAllCheck);
     MAKEATOM(funnel);

--- a/ecl/hqlcpp/hqlcatom.hpp
+++ b/ecl/hqlcpp/hqlcatom.hpp
@@ -50,6 +50,8 @@ extern IAtom * dictionaryAtom;
 extern IAtom * eclAtom;
 extern IAtom * endAtom;
 extern IAtom * ensureCapacityAtom;
+extern IAtom * externAtom;
+extern IAtom * externCAtom;
 extern IAtom * fileAtom;
 extern IAtom * forceAllCheckAtom;
 extern IAtom * funnelAtom;

--- a/ecl/hqlcpp/hqlcpp.ipp
+++ b/ecl/hqlcpp/hqlcpp.ipp
@@ -20,6 +20,7 @@
 #include <stdio.h>
 
 #include "jfile.hpp"
+#include "jexcept.hpp"
 #include "hqlattr.hpp"
 #include "hqlcpp.hpp"
 #include "hqlstmt.ipp"
@@ -463,6 +464,7 @@ public:
     unsigned numStmts() const;
     void setIncomplete(bool value);
     void setIncluded(bool value);
+    void addOption(IAtom * name) { dbgassertex(stmt); stmt->addOption(name); }
 
     inline bool isExecutedOnce() const { return (flags & MFsingle) != 0; }
 

--- a/ecl/hqlcpp/hqlhtcpp.cpp
+++ b/ecl/hqlcpp/hqlhtcpp.cpp
@@ -2461,8 +2461,9 @@ void ActivityInstance::buildSuffix()
             addConstructorMetaParameter();
 
             StringBuffer s;
-            s.append("extern \"C\" ECL_API IHThorArg * ").append(factoryName).append("()");
-            globalctx.addQuotedFunction(s, true);
+            s.append("IHThorArg * ").append(factoryName).append("()");
+            IHqlStmt * stmt = globalctx.addQuotedFunction(s, true);
+            stmt->addOption(externCAtom);
             OwnedHqlExpr call = translator.bindFunctionCall(implementationClassName, constructorArgs);
 
             //Don't call buildReturn because we're lying about the return type, and we don't want a boolean temporary created.
@@ -11238,17 +11239,18 @@ IWUResult * HqlCppTranslator::createDatasetResultSchema(IHqlExpression * sequenc
 
             BuildCtx transformctx(*code, declareAtom);
             transformctx.setNextPriority(RecordTranslatorPrio);
-            s.clear().append("extern \"C\" ECL_API size32_t ").append(name).append("(ARowBuilder & crSelf, const byte * src)");
+            s.clear().append("size32_t ").append(name).append("(ARowBuilder & crSelf, const byte * src)");
 
-            MemberFunction tranformFunc(*this, transformctx, s, MFdynamicproto);
+            MemberFunction transformFunc(*this, transformctx, s, MFdynamicproto);
+            transformFunc.addOption(externCAtom);
 
-            BoundRow * selfCursor = bindSelf(tranformFunc.ctx, tds, "crSelf");
-            bindTableCursor(tranformFunc.ctx, ds, "src", no_left, seq);
-            associateSkipReturnMarker(tranformFunc.ctx, queryZero(), selfCursor);
-            ensureRowAllocated(tranformFunc.ctx, "crSelf");
+            BoundRow * selfCursor = bindSelf(transformFunc.ctx, tds, "crSelf");
+            bindTableCursor(transformFunc.ctx, ds, "src", no_left, seq);
+            associateSkipReturnMarker(transformFunc.ctx, queryZero(), selfCursor);
+            ensureRowAllocated(transformFunc.ctx, "crSelf");
 
-            doTransform(tranformFunc.ctx, transform, selfCursor);
-            buildReturnRecordSize(tranformFunc.ctx, selfCursor);
+            doTransform(transformFunc.ctx, transform, selfCursor);
+            buildReturnRecordSize(transformFunc.ctx, selfCursor);
 
             result->setResultTransformerEntry(name.str());
         }

--- a/ecl/hqlcpp/hqlstmt.cpp
+++ b/ecl/hqlcpp/hqlstmt.cpp
@@ -1226,6 +1226,11 @@ void HqlStmt::addExpr(IHqlExpression * expr)
     exprs.append(*expr);
 }
 
+void HqlStmt::addOption(IAtom * name)
+{
+    addExpr(createAttribute(name));
+}
+
 StmtKind HqlStmt::getStmt() const
 {
     return (StmtKind)kind;
@@ -1251,6 +1256,10 @@ static bool isEmptyGroup(IHqlStmt * stmt)
 
 StringBuffer & HqlStmt::appendTextPrefix(StringBuffer & out) const
 {
+    if (hasAttribute(externAtom, exprs))
+        out.append("extern ");
+    if (hasAttribute(externCAtom, exprs))
+        out.append("extern \"C\" ECL_API ");
     if (noOptimize)
         out.append("NOOPTIMIZE ");
     if (optimize)

--- a/ecl/hqlcpp/hqlstmt.hpp
+++ b/ecl/hqlcpp/hqlstmt.hpp
@@ -228,6 +228,7 @@ public:
     virtual void            setIncomplete(bool incomplete) = 0;
     virtual void            setIncluded(bool _included) = 0;
     virtual void            finishedFramework() = 0;
+    virtual void            addOption(IAtom * name) = 0;
 };
 
 class HqlCppTranslator;

--- a/ecl/hqlcpp/hqlstmt.ipp
+++ b/ecl/hqlcpp/hqlstmt.ipp
@@ -43,6 +43,7 @@ public:
             bool                    isIncomplete()  { return incomplete; }
             unsigned                queryPriority() { return priority; }
     virtual void                    mergeScopeWithContainer()  override {}
+    virtual void                    addOption(IAtom * name) override;
     virtual void                    setIncomplete(bool _incomplete) override { incomplete = _incomplete; }
     virtual void                    setIncluded(bool _included) override { included = _included; }
     virtual void                    setForceOptimize(bool value) override;

--- a/ecl/hqlcpp/hqlwcpp.cpp
+++ b/ecl/hqlcpp/hqlwcpp.cpp
@@ -1751,7 +1751,7 @@ void HqlCppWriter::generateStmt(IHqlStmt * stmt)
                     indent(-1);
                     indent().append("}");
                     IHqlExpression * extra = stmt->queryExpr(0);
-                    if (extra)
+                    if (extra && !extra->isAttribute())
                         generateExprCpp(extra);
                     newline();
                 }

